### ISSUE TITLE
Tpetra: Replace Node with Kokkos space

### ISSUE THREAD

**Issue by mhoemmen** [trilinos/Trilinos#57]
@trilinos/tpetra 

This depends on #56 and #505.

Get rid of Node entirely.  Replace with Kokkos space.

**Comment by mhoemmen** [comment]
This depends on Issue #56.

**Comment by mhoemmen** [comment]
This issue supersedes Bugzilla Bug 6434

https://software.sandia.gov/bugzilla/show_bug.cgi?id=6434

If you can't access the above link, here is the original text (writted 12 Oct 2015):

Currently, Tpetra classes specify what Kokkos execution space and memory space to use via the Node template parameter.  Tpetra classes currently require that Node be a specialization of Kokkos::Compat::KokkosDeviceWrapperNode<ExecutionType, MemorySpaceType>, where ExecutionType is a Kokkos execution space, and MemorySpaceType is a Kokkos memory space.

Kokkos::Compat::KokkosDeviceWrapperNode was a backwards compatibility measure to aid porting Tpetra from the "classic" version to the "Kokkos refactor" version.  Now that the "classic" version of Tpetra is gone, we have no need for the classic "Node API."  Furthermore, it is confusing to users.  Users who already know about Kokkos would prefer to put a Kokkos::Device specialization right into Tpetra as a template parameter, in order to specify what execution space and memory space Tpetra should use.

We want to do the following:

  1. Stop using the Node interface altogether
  2. Let Tpetra users give a Kokkos::Device specialization as a Tpetra template parameter
  3. Deprecate and remove Kokkos::Compat::KokkosDeviceWrapperNode.

Tpetra requires C++11, so we can use C++11 type alias templates to simplify this process and maintain backwards compatibility, without adding more ETI instantiations.  For example, we could make Kokkos::Compat::KokkosDeviceWrapperNode an alias for Kokkos::Device (it takes the same template parameters).

I'm only adding this bug now because it is taking up space on my white board and I need to record it.

**Comment by mhoemmen** [comment]
See also #1889.  The fix for that issue will make Nodes even less important.  Eventually, we can have Map take over the duties of Node, with respect to Kokkos initialization and finalization (if Kokkos hasn't already been initialized).

**Comment by mhoemmen** [comment]
> Eventually, we can have Map take over the duties of Node, with respect to Kokkos initialization and finalization (if Kokkos hasn't already been initialized).

PR #3044, recently merged into develop, makes Tpetra::Map take over the Kokkos initialization and finalization duties of Node.  See also PR #3058, that moves this initialization code out of Teuchos and into Tpetra.

**Comment by mhoemmen** [comment]
PR #3060, just merged into develop, makes progress on this issue.  Here is a summary:

1. Xpetra,Tpetra: Purge `defaultArgNode`; just create a Node instance
2. Tpetra: Remove `GO=int` dependency from Bug 5401 test
3. Tpetra: Deprecate `KokkosClassic::Details::get(Default)Node`
4. Xpetra: `create(Uniform){Local,Contig}MapWithNode` Node argument is now optional
5. Purge use of `getNode` from TpetraCore and Belos tests
6. Fix build warnings and clean up many tests
